packages/jsii-pacmak/lib/targets/golang/module.ts

```diff
@@ -1,64 +1,254 @@
 import { CodeMaker } from 'codemaker';
-import { Assembly, Submodule as JsiiSubmodule } from 'jsii-reflect';
+import { Assembly, Type, Submodule as JsiiSubmodule } from 'jsii-reflect';
+import { join } from 'path';
 import { GoClass, Enum, Interface } from './types';
 
-export interface ModuleTypes {
-  [fqn: string]: Interface | Enum | GoClass;
+type ModuleType = Interface | Enum | GoClass;
+type ModuleTypes = ModuleType[];
+
+// TODO: Make this a class?
+export interface Module {
+  root: RootModule;
+  packageName: string;
+  moduleName: string;
+  filePath: string;
+  file: string;
+  submodules: Submodule[];
+  types: ModuleTypes;
+  hasSubmodules(): boolean;
+}
+
+function buildSubmodules(
+  root: RootModule,
+  parent: Module,
+  submodules: readonly JsiiSubmodule[],
+): Submodule[] {
+  return submodules.map((sm) => new Submodule(root, parent, sm));
 }
```

Comment:
I don't like this. This is extremely wasteful - since this is invoked by the `.submodules` property of anything module-like (and your unnecessary `hasSubmodules` accessor actually checks the length of that), the result of this should **at the very least** be cached.

But it looks like this could be folded in a "module-like" base class, which would simply compute this at construction time and hold on to it forever after.

Reply:
Yeah I was having trouble getting the module base class to look the way I wanted it to (mostly it had a crappy constructor signature with a ton of args). I agree though and I'll take another crack.